BREAKING: Simplify foreign keys that are also primary keys

hgl opened the pull request:
## Description

<!-- If this PR adds a feature, what does it add and what was the motivation for it? -->
<!-- If this PR fixes an issue, what is the issue? -->
<!-- Please link to the relevant GitHub issues/Discord discussion/etc as appropriate -->
Fixes #34

One unsolved issue is that I think I'm also supposed to fix `manyRelationByKeysSimple`, but it's never called for the `animal` test, and there is no `dogByAnimalIdList` simple connection generated for the unsimplified case. Is it broken by my change or related to another plugin?
## Performance impact
unknown
<!-- Detail any impact on performance of this PR, or put 'unknown' if not known -->

## Security impact
unknown
<!-- Detail any impact on security of this PR, or put 'unknown' if not known -->

## Checklist

<!-- If this PR is work in progress, please open it as a "Draft PR". -->
<!-- To tick a checkbox, change it from `[ ]` to `[x]` -->

- [x] My code matches the project's code style and `yarn lint:fix` passes.
- [x] I've added tests for the new feature, and `yarn test` passes.
- [ ] I have detailed the new feature in the relevant documentation.
- [ ] I have added this feature to 'Pending' in the `RELEASE_NOTES.md` file (if one exists).
- [ ] If this is a breaking change I've explained why.

<!-- For some Graphile projects the documentation is the README.md file, for
      others please see https://github.com/graphile/graphile.github.io -->

hgl commented:
Ah, never mind, since primary keys are unique, there won't be connections or lists in the backward relations. I think it's good for review @benjie.

benjie commented on index.js:
```diff
@@ -195,24 +195,29 @@ function PgSimplifyInflectorPlugin(
           return this.camelCase(baseName);
         }
-        if (this.baseNameMatches(baseName, table.name)) {
+        if (this._isPrimaryKey(detailedKeys, foreignTable)) {
```
Please can you restore the `baseNameMatches` call, or explain why it has been removed?

hgl replied:
If I'm not wrong, it's not possible for this condition to be true, because the previous construct is

```
if (baseName) {
          return this.camelCase(baseName);
}
```

So `baseName` must be falsey here. From the implementation of `baseNameMatches`, it shouldn't be possible for `table.name` to match something falsey.

hgl replied:
The refactoring of `oppositeBaseName` and friends follows the same logic.

benjie replied:
Good catch :+1:

benjie commented on index.js:
```diff
@@ -195,24 +195,29 @@ function PgSimplifyInflectorPlugin(
           constraint
         );
       },
+      _isPrimaryKey(detailedKeys, table) {
+        return detailedKeys.every(
+          (key, i) => key === table.primaryKeyConstraint.keyAttributes[i]
+        );
+      },
```
This is not an inflector; please extract it out of the inflectors object to the top scope as a utility function.

hgl replied:
Fixed.

benjie commented on index.js:
```diff
@@ -195,24 +195,29 @@ function PgSimplifyInflectorPlugin(
         );
       },
+      _isPrimaryKey(detailedKeys, table) {
+        return detailedKeys.every(
+          (key, i) => key === table.primaryKeyConstraint.keyAttributes[i]
+        );
```
```suggestion
        return detailedKeys.length === table.primaryKeyConstraint.keyAttributes.length && detailedKeys.every(
          (key, i) => key === table.primaryKeyConstraint.keyAttributes[i]
        );
```

hgl replied:
Good catch.

benjie commented on tests/animal/schema.sql:
```diff
@@ -0,0 +1,7 @@
+  id serial primary key
+);
+
+create table dog (
+  id integer primary key references animal
+);
```
Lets add a few more examples to see that sensible behaviour continues even if there are many relationships.

```suggestion
create table cat (
  id integer primary key references animal
);

create table dog (
  id integer primary key references animal
);

create table gerbil (
  animal_id integer primary key references animal
);
```

hgl replied:
Added.

benjie reviewed (changes requested):
Looking good! Can you think of more edge cases we might need to test? What sort of SQL schemas would cause this to trigger naming conflicts?

hgl commented:
I just realized I edited on `index.js` instead of `index.ts`. It's now updated. Also fixed a couple type errors.

> What sort of SQL schemas would cause this to trigger naming conflicts?

This change simply names the property the same table name that the corresponding primary key points to. So unless there is an existing property with that name, there shouldn't be such conflict, and property name conflicts should be considered an user error? Let me know if I missed anything.

hgl commented:
Also, do you think it’s a good idea to not commit index.js in the repo?

benjie commented:
> Also, do you think it’s a good idea to not commit index.js in the repo?

For this plugin specifically it allows you to use the repo directly rather than the npm module. I only allow it here because this plugin is so simple (one single file).

> I just realized I edited on index.js instead of index.ts.

I was about to let you know this myself; good catch!

> > What sort of SQL schemas would cause this to trigger naming conflicts?

> This change simply names the property the same table name that the corresponding primary key points to. So unless there is an existing property with that name, there shouldn't be such conflict, and property name conflicts should be considered an user error? Let me know if I missed anything.

I like to test the edges of things, see where they'll fall over. So for example if you have multiple relationships between records:

```sql
create table comments (
  id serial primary key
);
create table comment_extras (
  id serial primary key references comments,
  parent_comment_id int references comments
);
```

In this case it should give you

```graphql
type CommentExtras {
  comment: Comment
  parentComment: Comment
}
```

I'm wondering if there's a common use case when you might reference the target again with what appears to be the same name e.g. what would valid values for "foo" and "bar" be in the following example? I can't think of any...

```sql
create table foo (
  id serial primary key
);
create table bar (
  id serial primary key references foo,
  foo_id int references foo
);
```

benjie commented:
This looks good to me, but I'm too exhausted to trust my review right now so I'm going to re-review it on another day. I don't think any more changes are required though :+1:

hgl commented:
> I like to test the edges of things, see where they'll fall over

That's very thoughtful. I think I have the same kind of mentality when writing my own code, but the bar might be lowered when contributing to unfamiliar code 😂. Thanks for the reminding.

> what would valid values for "foo" and "bar" be in the following example

I tried it out actually, and since the FK is not unique, it just generates an `fooId` property. The conflict only happens when you give the FK an unique constraint.

I can't really come up with a reasonable example where the primary key and an unique FK points to the same table, and the FK has to be named as table name (or + `_id`). The latter seems to always suggest redundancy.

But I think this might be a good reminder that the docs could probably provide some guidance on polymorphism table design. I sent this PR because after experimenting, I landed on this design that I like the most:

```sql
create table animal (
  id serial primary key
);
create table cat (
  id integer primary key references animal,
);
create table dog (
  id integer primary key references animal,
);
```
This is creates both the cleanest tables and GraphQL types, imho, comparing to other designs like:

```sql
create table animal (
  id serial primary key,
  cat_id integer unique references cat,
  dog_id integer unique references dog
);
```
or
```sql
create table cat (
  id serial primary key,
  animal_id integer unique references animal
);
```
Which either feels messy or redundant.

If the official doc could provide some guidance on this and similar topics, maybe many edge cases (at least ones result from bad designs) could be avoided.

---

Anyway, thanks for taking the time reviewing this.

benjie commented:
> If the official doc could provide some guidance on this and similar topics, maybe many edge cases (at least ones result from bad designs) could be avoided.

Agreed; this is something I hope to address in V5, so I don't want to document a pattern that might change or require tweaks just yet. Keep in mind this PR isn't limited to polymorphism though, just anything that has additional data. (e.g. you could have an RPG database with `characters` and `character_stats` (strength, dexterity, etc), `character_health` (hp, armour, etc), `character_buffs` (is_strengthened, is_stunned, is_frozen, is_burning, etc) etc - in this case you might design it with lots of tables that each have a primary key that points to the `characters` table.)

> I can't really come up with a reasonable example where the primary key and an unique FK points to the same table, and the FK has to be named as table name (or + _id). The latter seems to always suggest redundancy.

Ah; how about you have a system that encourages you to take a book as inspiration and write a new book based on it. (Contrived, I know...) So first you create entries into `ideas` with the books that are your inspiration, and then as you write the derivative books you set the `book_id` to point at your final book. e.g.

```sql
create table books (
  id serial primary key
);
create table ideas (
  inspiration_book_id serial primary key references books,
  book_id int unique references books
);
```

I'd want this to turn into

```graphql
type Idea {
  inspirationBook: Book! # ideally non-null, but would be nullable by default
  book: Book 
}
```

Would it cause a conflict currently?

hgl commented:
> Would it cause a conflict currently?

Yes it would, but in a very interesting way:

For forward keys, `Idea` would turn into what you'd expected, because the  `this.baseNameMatches` test would pass, before the primary key test has the chance (this PR only takes effect when the primary key fails the base name match).

For backward keys, `inspiration_book_id` generates `book.idea`. But so does `book_id`, because the base name `book` matches the foreign table name, and it's an unique constraint.

I'd argue that for backward keys, `book.idea` being generated from `ideas`' primary key is more intuitive than from an unique foreign key, and I think giving a FK key a primary-key-like name qualifies as a bad design. You can avoid this conflict by changing `book_id` to `derived_book_id` for example, which generates `book.derivedBook`.

(The forced push is to add the missing code for `singleRelationByKeys` while converting from js to ts.)

benjie commented:
Yeah I agree. I'm happy we've considered this :+1:

benjie reviewed (approved):
I found a minor issue which I've fixed but I think we're good to go now :+1:

benjie commented:
This is released in @graphile-contrib/pg-simplify-inflector@next (v7.0.0-alpha.1)

hgl commented:
Thanks for the fix.
